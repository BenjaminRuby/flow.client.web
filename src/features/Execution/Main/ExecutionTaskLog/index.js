import React, { useState } from "react";
import PropTypes from "prop-types";
import orderBy from "lodash/orderBy";
import { Arrows32, ChevronLeft32 } from "@carbon/icons-react";
import TaskItem from "./TaskItem";
import { ACTIVITY_STATUSES_TO_ICON, ACTIVITY_STATUSES_TO_TEXT } from "Constants/activityStatuses";
import { getSimplifiedDuration } from "Utilities/timeHelper";
import styles from "./executionTaskLog.module.scss";

ExecutionTaskLog.propTypes = {
  workflowExecutionData: PropTypes.object.isRequired
};

function ExecutionTaskLog({ workflowExecutionData }) {
  const [isCollapsed, setIsCollapsed] = useState(false);
  const [tasksSort, setTasksSort] = useState("desc");

  const { id, duration, status, steps } = workflowExecutionData;
  const Icon = ACTIVITY_STATUSES_TO_ICON[status];

  const toggleCollapse = () => {
    setIsCollapsed(!isCollapsed);
  };

  const toggleSort = () => {
    setTasksSort(tasksSort === "desc" ? "asc" : "desc");
  };

  const sortedTasks = orderBy(steps, step => step.order, [tasksSort]);

  return (
    <aside className={`${styles.container} ${isCollapsed ? styles.collapsed : ""}`}>
      <section className={`${styles.statusBlock} ${styles[status]}`}>
        <div className={styles.duration}>
          <p className={styles.title}>Duration</p>
          <time className={styles.value}>
            {typeof duration === "number" ? getSimplifiedDuration(duration / 1000) : "--"}
          </time>
        </div>
        <div className={styles.status}>
          <p className={styles.title}>Status</p>
          <div className={styles.statusData}>
            <Icon aria-label={status} className={styles.statusIcon} />
            <p className={styles.value}>{ACTIVITY_STATUSES_TO_TEXT[status]}</p>
          </div>
        </div>
        <button className={styles.collapseButton} onClick={toggleCollapse}>
          <ChevronLeft32 className={styles.chevron} />
        </button>
      </section>
      <section className={styles.taskbar}>
        <p className={styles.taskbarTitle}>Task log</p>
        <button className={styles.taskbarButton} onClick={toggleSort}>
          <Arrows32 className={styles.taskbarArrows} />
        </button>
      </section>
      <ul className={styles.tasklog}>
        {sortedTasks.map(step => (
          <TaskItem flowActivityId={id} task={step} />
        ))}
      </ul>
    </aside>
  );
}

export default ExecutionTaskLog;
