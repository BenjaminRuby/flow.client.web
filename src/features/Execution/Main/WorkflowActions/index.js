import React from "react";
import PropTypes from "prop-types";
import { withRouter } from "react-router-dom";
import { Edit32 } from "@carbon/icons-react";
import styles from "./WorkflowActions.module.scss";

WorkflowActions.propTypes = {
  history: PropTypes.object.isRequired,
  setActiveTeam: PropTypes.func.isRequired,
  workflow: PropTypes.object.isRequired
};

function WorkflowActions(props) {
  function setActiveTeamAndRedirect() {
    const { id, flowTeamId } = props.workflow;

    props.setActiveTeam(flowTeamId);
    props.history.push(`/editor/${id}/designer`);
  }
  return (
    <div className={styles.container}>
      <p className={styles.messageText}>Read-only</p>
      <button className={styles.editContainer} onClick={setActiveTeamAndRedirect}>
        <Edit32 className={styles.editIcon} />
        <p className={styles.editText}>Edit Workflow</p>
      </button>
    </div>
  );
}

export default withRouter(WorkflowActions);
