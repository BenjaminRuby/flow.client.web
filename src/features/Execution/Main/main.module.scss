@import "~Styles/variables";

.container {
  background-color: $ui-background !important;
  height: 100%;
  width: 100%;
}

.executionResultContainer {
  display: flex;
  height: calc(100vh - 10.3125rem);
}

.executionDesignerContainer {
  display: flex;
  flex-grow: 1;
  position: relative;

  svg:not(:root) {
    overflow: visible !important;
  }
}

.diagram {
  margin-top: 3.5rem;
}

.executionWorkflowActions {
  background-color: $ui-background;
  border-bottom: 0.0625rem solid #b9bfc7;
  display: flex;
  justify-content: space-between;
  height: 3.5rem;
  position: absolute;
  width: 100%;
}
