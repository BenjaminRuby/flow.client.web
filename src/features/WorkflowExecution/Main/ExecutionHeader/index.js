import React from "react";
import PropTypes from "prop-types";
import { withRouter, Link } from "react-router-dom";
import moment from "moment";
import styles from "./executionHeader.module.scss";

ExecutionHeader.propTypes = {
  workflowExecutionData: PropTypes.object.isRequired
};

function ExecutionHeader({ history, workflow, workflowExecutionData }) {
  const { state } = history.location;

  const data = {
    Team: workflowExecutionData.teamName,
    "Initiated by": workflowExecutionData.initiatedByUserName,
    Trigger: workflowExecutionData.trigger,
    "Start time": moment(workflowExecutionData.creationDate).format("YYYY-MM-DD hh:mm A")
  };

  return (
    <header className={styles.container}>
      <div className={styles.headline}>
        <div className={styles.subtitle}>
          <Link to={state ? state.fromUrl : "/activity"}>{state ? state.fromText : "Activity"}</Link>
          <p style={{ margin: "0 0.5rem" }}>/</p>
          <p>{workflow.name}</p>
        </div>
        <p className={styles.title}>Workflow run detail</p>
      </div>
      <div className={styles.content}>
        {Object.keys(data).map(key => (
          <div key={key} className={styles.data}>
            <p className={styles.dataTitle}>{key}</p>
            <p className={styles.dataValue}>{data[key]}</p>
          </div>
        ))}
      </div>
    </header>
  );
}

export default withRouter(ExecutionHeader);
