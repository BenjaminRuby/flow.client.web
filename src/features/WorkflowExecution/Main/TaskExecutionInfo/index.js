import React from "react";
import PropTypes from "prop-types";
import moment from "moment";
import getHumanizedDuration from "@boomerang/boomerang-utilities/lib/getHumanizedDuration";
import TaskExecutionLog from "./TaskExecutionLog";
import OutputPropertiesLog from "./OutputPropertiesLog";
import { ACTIVITY_STATUSES_TO_TEXT, ACTIVITY_STATUSES_TO_ICON } from "Constants/activityStatuses";
import "./styles.scss";

const TaskExecutionInfo = ({ flowActivityId, task }) => {
  return (
    <aside className="c-step-side-info">
      <h1 className="s-step-side-info-title">{task.taskName}</h1>
      <section className="c-step-side-info-details">
        <dl className="b-step-side-info-field">
          <dt className="b-step-side-info-field__key">Type</dt>
          <dd className="b-step-side-info-field__value">{"Task"}</dd>
        </dl>
        <dl className="b-step-side-info-field">
          <dt className="b-step-side-info-field__key">Status</dt>
          <dd className="b-step-side-info-field__value">
            {ACTIVITY_STATUSES_TO_TEXT[task.flowTaskStatus]}
            <img src={ACTIVITY_STATUSES_TO_ICON[task.flowTaskStatus]} alt="Status" />
          </dd>
        </dl>
        <dl className="b-step-side-info-field">
          <dt className="b-step-side-info-field__key">Start Time</dt>
          <dd className="b-step-side-info-field__value">{moment(task.startTime).format("MMMM Do YYYY, h:mm:ss a")}</dd>
        </dl>
        <dl className="b-step-side-info-field">
          <dt className="b-step-side-info-field__key">Duration</dt>
          <dd className="b-step-side-info-field__value">
            {task.duration ? getHumanizedDuration(Math.round(parseInt(task.duration / 1000), 10)) : "---"}
          </dd>
        </dl>
        <dl className="b-step-side-info-field">
          <dt className="b-step-side-info-field__key">Log</dt>
          <dd className="b-step-side-info-field__value">
            <TaskExecutionLog
              flowTaskStatus={task.flowTaskStatus}
              flowTaskId={task.taskId}
              flowActivityId={flowActivityId}
              flowTaskName={task.taskName}
            />
          </dd>
        </dl>
        {task.outputs && Object.keys(task.outputs).length > 0 && (
          <dl className="b-step-side-info-field">
            <dt className="b-step-side-info-field__key">Properties</dt>
            <dd className="b-step-side-info-field__value">
              <OutputPropertiesLog flowTaskName={task.taskName} flowTaskOutputs={task.outputs} />
            </dd>
          </dl>
        )}
      </section>
    </aside>
  );
};

TaskExecutionInfo.propTypes = {
  flowActivityId: PropTypes.string.isRequired,
  task: PropTypes.object.isRequired
};

export default TaskExecutionInfo;
