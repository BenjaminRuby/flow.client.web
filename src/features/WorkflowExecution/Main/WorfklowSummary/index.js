import React from "react";
import PropTypes from "prop-types";
import moment from "moment";
import { ACTIVITY_STATUSES_TO_TEXT, ACTIVITY_STATUSES_TO_ICON } from "Constants/activityStatuses";
import getHumanizedDuration from "@boomerang/boomerang-utilities/lib/getHumanizedDuration";
import "./styles.scss";

const WorkflowSummary = ({ workflowData, workflowExecutionData, version }) => {
  return (
    <aside className="c-workflow-summary">
      <h1 className="s-workflow-summary-title">{workflowData.name || "Workflow"}</h1>
      <section className="c-workflow-summary-details">
        <dl className="b-workflow-summary-field">
          <dt className="b-workflow-summary-field__key">Summary</dt>
          <dd className="b-workflow-summary-field__value">{workflowData.shortDescription || "---"}</dd>
        </dl>
        <dl className="b-workflow-summary-field">
          <dt className="b-workflow-summary-field__key">Description</dt>
          <dd className="b-workflow-summary-field__value">{workflowData.description || "---"}</dd>
        </dl>
        <dl className="b-workflow-summary-field">
          <dt className="b-workflow-summary-field__key">Version</dt>
          <dd className="b-workflow-summary-field__value">{version}</dd>
        </dl>
        <dl className="b-workflow-summary-field">
          <dt className="b-workflow-summary-field__key">Status</dt>
          <dd className="b-workflow-summary-field__value">
            {ACTIVITY_STATUSES_TO_TEXT[workflowExecutionData.status ? workflowExecutionData.status : "notstarted"]}
            <img
              className="b-activity-card__status-icon"
              src={
                ACTIVITY_STATUSES_TO_ICON[workflowExecutionData.status ? workflowExecutionData.status : "notstarted"]
              }
              alt={`Status ${workflowExecutionData.status}`}
            />
          </dd>
        </dl>
        <dl className="b-workflow-summary-field">
          <dt className="b-workflow-summary-field__key">Start Time</dt>
          <dd className="b-workflow-summary-field__value">
            {moment(workflowExecutionData.creationDate).format("MMMM Do YYYY, h:mm:ss a")}
          </dd>
        </dl>
        <dl className="b-workflow-summary-field">
          <dt className="b-workflow-summary-field__key">Duration</dt>
          <dd className="b-workflow-summary-field__value">
            {workflowExecutionData.duration
              ? getHumanizedDuration(parseInt(workflowExecutionData.duration / 1000, 10))
              : "---"}
          </dd>
        </dl>
      </section>
    </aside>
  );
};

WorkflowSummary.propTypes = {
  workflowData: PropTypes.object.isRequired,
  workflowExecutionData: PropTypes.object.isRequired,
  version: PropTypes.number.isRequired
};

export default WorkflowSummary;
