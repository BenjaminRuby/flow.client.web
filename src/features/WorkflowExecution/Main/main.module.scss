@import "~Styles/variables";

.container {
  background-color: $ui-background !important;
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  position: relative;
  width: 100%;
}

.executionResultContainer {
  display: flex;
  flex-grow: 1;
}

.executionDesignerContainer {
  display: flex;
  flex-grow: 1;
  position: relative;
}
