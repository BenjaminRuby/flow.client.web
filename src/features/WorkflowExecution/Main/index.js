import React, { Component } from "react";
import PropTypes from "prop-types";
import { withRouter } from "react-router-dom";
import { DiagramWidget } from "@projectstorm/react-diagrams";
import { LoadingAnimation } from "@boomerang/carbon-addons-boomerang-react";
import ExecutionHeader from "./ExecutionHeader";
import ExecutionTaskLog from "./ExecutionTaskLog";
import DiagramApplication from "Utilities/DiagramApplication";
import { EXECUTION_STATUSES } from "Constants/workflowExecutionStatuses";
import styles from "./main.module.scss";

class Main extends Component {
  static propTypes = {
    dag: PropTypes.object.isRequired,
    workflowData: PropTypes.object.isRequired,
    workflowExecutionData: PropTypes.object.isRequired
  };

  constructor(props) {
    super(props);
    this.diagramApp = new DiagramApplication({ dag: props.dag, modelIsLocked: true });
  }

  render() {
    const { workflowData, workflowExecutionData, setActiveTeam } = this.props;
    const hasFinished = [EXECUTION_STATUSES.COMPLETED, EXECUTION_STATUSES.INVALID, EXECUTION_STATUSES.FAILURE].includes(
      workflowExecutionData.status
    );

    const hasStarted =
      workflowExecutionData.steps &&
      workflowExecutionData.steps.find(step => step.flowTaskStatus !== EXECUTION_STATUSES.NOT_STARTED);

    return (
      <div className={styles.container}>
        <ExecutionHeader workflow={workflowData} workflowExecutionData={workflowExecutionData} />
        <ExecutionTaskLog
          setActiveTeam={setActiveTeam}
          workflow={workflowData}
          workflowExecutionData={workflowExecutionData}
        />
        {hasStarted || hasFinished ? (
          <DiagramWidget
            allowLooseLinks={false}
            allowCanvasTranslation={true}
            allowCanvasZoom={true}
            className={styles.diagram}
            deleteKeys={[]}
            diagramEngine={this.diagramApp.getDiagramEngine()}
            maxNumberPointsPerLink={0}
          />
        ) : (
          <LoadingAnimation message="Your workflow will be with you shortly" />
        )}
      </div>
    );
  }
}

export default withRouter(Main);
