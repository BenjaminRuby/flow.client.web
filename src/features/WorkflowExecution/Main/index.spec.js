import React from "react";
import Main from "./index";

jest.mock("@boomerang/boomerang-components", () => ({
  LoadingAnimation: "LoadingAnimation"
}));

const mockfn = jest.fn();

const props = {
  dag: {},
  taskId: "test",
  workflowData: {},
  workflowExecutionData: {},
  version: 1,
  activityState: {
    status: "success"
  },
  teamsState: {
    status: "success"
  },
  updateActiveNode: mockfn
};

describe("Main --- Snapshot", () => {
  it("Capturing Snapshot of Main", () => {
    const { baseElement } = rtlRouterRender(<Main {...props} />);
    expect(baseElement).toMatchSnapshot();
  });
});
