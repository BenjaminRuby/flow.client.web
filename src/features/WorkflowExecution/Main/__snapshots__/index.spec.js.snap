// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main --- Snapshot Capturing Snapshot of Main 1`] = `
<body
  id="app"
>
  <div>
    <div
      class="container"
    >
      <header
        class="container border"
      >
        <div
          class="container"
        >
          <section>
            <div
              class="subtitle"
            >
              <a
                href="/activity"
              >
                Activity
              </a>
              <p
                style="margin: 0px 0.5rem;"
              >
                /
              </p>
              <p>
                Sparkle Flow with extra glitter and donuts on the side
              </p>
            </div>
            <h1
              class="title"
            >
              Workflow run detail
            </h1>
          </section>
          <div
            class="content"
          >
            <div
              class="data"
            >
              <p
                class="dataTitle"
              >
                Team
              </p>
              <p
                class="dataValue"
              />
            </div>
            <div
              class="data"
            >
              <p
                class="dataTitle"
              >
                Initiated by
              </p>
              <span
                aria-hidden="false"
                aria-label="robot"
                role="img"
              >
                ðŸ¤–
              </span>
            </div>
            <div
              class="data"
            >
              <p
                class="dataTitle"
              >
                Trigger
              </p>
              <p
                class="dataValue"
              />
            </div>
            <div
              class="data"
            >
              <p
                class="dataTitle"
              >
                Team
              </p>
              <p
                class="dataValue"
              >
                2019-01-01 12:00 AM
              </p>
            </div>
          </div>
        </div>
      </header>
      <main
        class="executionResultContainer"
      >
        <aside
          class="container "
        >
          <section
            class="statusBlock inProgress"
          >
            <div
              class="duration"
            >
              <p
                class="title"
              >
                Duration
              </p>
              <time
                class="value"
              >
                --
              </time>
            </div>
            <div
              class="status"
            >
              <p
                class="title"
              >
                Status
              </p>
              <div
                class="statusData"
              >
                <svg
                  aria-label="inProgress"
                  class="statusIcon"
                  focusable="false"
                  height="32"
                  preserveAspectRatio="xMidYMid meet"
                  role="img"
                  style="will-change: transform;"
                  viewBox="0 0 32 32"
                  width="32"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M15 11h2v9h-2zm-2-9h6v2h-6z"
                  />
                  <path
                    d="M28 9l-1.42-1.41-2.25 2.25a10.94 10.94 0 1 0 1.18 1.65zM16 26a9 9 0 1 1 9-9 9 9 0 0 1-9 9z"
                  />
                </svg>
                <p
                  class="value"
                >
                  In Progress
                </p>
              </div>
            </div>
            <button
              class="collapseButton"
            >
              <svg
                aria-hidden="true"
                class="chevron"
                focusable="false"
                height="32"
                preserveAspectRatio="xMidYMid meet"
                style="will-change: transform;"
                viewBox="0 0 32 32"
                width="32"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10 16L20 6l1.4 1.4-8.6 8.6 8.6 8.6L20 26z"
                />
              </svg>
            </button>
          </section>
          <section
            class="taskbar"
          >
            <p
              class="taskbarTitle"
            >
              Task log
            </p>
            <button
              class="taskbarButton"
            >
              <svg
                aria-hidden="true"
                class="taskbarArrows"
                focusable="false"
                height="32"
                preserveAspectRatio="xMidYMid meet"
                style="will-change: transform;"
                viewBox="0 0 32 32"
                width="32"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M26.6 20.6L23 24.2V4h-2v20.2l-3.6-3.6L16 22l6 6 6-6zM10 4l-6 6 1.4 1.4L9 7.8V28h2V7.8l3.6 3.6L16 10z"
                />
              </svg>
            </button>
          </section>
          <ul
            class="tasklog"
          />
        </aside>
      </main>
    </div>
  </div>
</body>
`;
